src/rust/engine/BUILD

```diff
@@ -0,0 +1,39 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+_RELEASE = env("MODE") == "release"
+_RUST_MODE = "--release" if _RELEASE else ""
+_OUTPUT_DIR = "release" if _RELEASE else "debug"
+
+files(
+    name="rust_sources",
+    sources=[
+        "Cargo.lock",
+        "build.rs",
+        "**/Cargo.toml",
+        "**/*.rs",
+        "**/*.proto",
+        ".cargo/config",
+    ],
+)
+
+shell_sources(name="shell")
+
+shell_command(
+    name="engine-and-client",
+    command=f"./sandboxed_cargo.sh build {_RUST_MODE} --features=extension-module -p engine -p client",
+    execution_dependencies=[
+        "./sandboxed_cargo.sh:shell",
+        ":rust_sources",
+        "3rdparty/tools/protoc:protoc",
+        "3rdparty/tools/python3:python3",
+        "3rdparty/tools/rust:rust-toolchain",
+    ],
+    extra_env_vars=["CHROOT={chroot}", "MODE"],
+    tools=["cc", "ld", "as", "ar"],
```

Comment:
Did you consider passing in cargo and rustc here too, and/or using `adhoc_tool` + `system_binary` like the earlier experiments?

That is, preserve the current behaviour of requiring the user to have installed Rustup, rather than duplicating rust-toolchain

(Having Rust all automatically managed seems cool, but also is potentially an improvement we can do as follow up, to be able to discuss the specific trade-offs?)

Reply:
Yeah, so I think it boils down to where we want the slider to be on "hermetic+reproducible" vs "simplicity"

Generally speaking if it's not too hard to get everything working in the hermetic/reproducible camp that's my default. It means that the build process is much safer to cache, and there'll be fewer flakes. 

But that's only if it's easy. So things like the system binaries required, which we pass as tools... Yeah that we can poke a hole in the sandbox for.


Honestly, I think Pants needs more/better tooling for upgrading/validating http_source objects. So that's on us to make easier 🙂

Reply:
As for adhoc_tool, not only was this shorter/simpler, but I also had a lot of issues with supporting Pants-provided tools that aren't single-file as binaries on PATH. So this was was much simpler.

Reply:
Sorry, I think my question was a bit unclear: I am fully on-board with fully hermetic eventually. I am more asking about the timing/phasing: this awesome work currently simultaneously:

1. switches us to have pants orchestrate the Rust build 🎉 
2. changes how we install Rust/Cargo, to be unconventional ❓ 

Doing 1 seems great, but it seems like 2 could potentially cause issues in the short term, and it would nice for the benefits of 1 _not_ be stuck behind working through any problems/debate with 2.

(I have an impression that running Rust manually can be a bit tricky, with concepts of "sysroots" and so on rattling around in my head (with scars...).)

Once 1 is in, then we can focus on landing 2.

Reply:
Hmm maybe I'm being unimaginative... Can you elaborate what you feel might be problematic? If anything this gives me more of the warm fuzzies because we've controlled exactly the pieces that went into compilation.

It also helps to me, that developers have a 0-ish install way of contributing. I'm thinking of that user who recently had so many issues getting started, and the recent headaches with protoc.

Reply:
I feel like we're talking past each other. At no point have I suggested that we _shouldn't_ do this, but you seem to be responding as if I am? As I said, I'm a big fan of using pants to make for 0-install contribution set-up too!

However, that feels like making this PR do too many things, and we should break up the independent parts for easier/faster review (and also easier/faster bisect when there's problems, e.g. when we inevitably make a mistake or don't realise some subtlety and a contributor comes to us saying "my setup is now broken", landing over several steps makes it much easier to work out exactly which part of this process broke it for them).

As in, I'm suggesting we do this in several steps:

1. start using pants to orchestrate Rust build without _changing_ what users need to install (so the installation requirements don't get _better_ (but also don't get worse), but we can start seeing the benefits of having pants run the Rust build), e.g. passing `tools=[..., "cargo", "rustc"]` to start with
2. as (close-after) follow up, use pants' powers to get to zero-install (now the installation requirements get better! 🎉 )

For the specifics of running Rust in an unconventional way, I have vague memories of having to fight "sysroots" and so on when I was contributing to it a lot, with rustup managing all of that smoothly (including things like "running Rust from source on repos"). This is many years ago, so I don't have specifics problems in mind, just being having a vague sense of "here be dragons" and "if we do the steps separately, we can land each step sooner, without getting tangled in unnecessary dragons".

Reply:
Sorry if you felt that way. I think I was just missing the core of the _why_. I also got very excited at the prospect of this change. To me, doing 1 isn't _that_ much of an improvement, so I missed the more nuanced aspects of it.

That being said, I can't say I still see if it's _that_ much more valuable to split it up, but happy to concede if that's the consensus.

Anything is better than the status quo. 🙂

Reply:
Just step 1 (i.e. pants is a normal pants repo) seems like something to get excited about too: e.g. not having to rebuild Rust to run a build support script, and pantsd being invalidated far less often. I also take from https://github.com/pantsbuild/pants/discussions/19774 that you think that step 1 is necessary for getting to smoother releases.

After that, I'm also very excited by the additional features to get to zero-install setup.

But, major infrastructure/fundamental changes like this are easy to get wrong (e.g. someone has a weird/different system), and I've been doing a lot of debugging of our other major infrastructure/fundamental changes in Pants recently, and it's pretty frustrating. Doing the changes in smaller/smoother steps will make that inevitable debugging easier.

Reply:
https://github.com/pantsbuild/pants/pull/19817